Allow unloading DNS policy rules on graceful shutdown

tklauser opened the pull request:
Introduce the `--dns-policy-unload-on-shutdown` option which will cause
DNS policy rules to be unloaded on graceful agent shutdown.
    
This leads to DNS queries no longer being redirected to the L7 DNS proxy
by the BPF datapath and allows endpoints on a node where the agent is
not running to continue being able to resolve DNS queries.
    
By default the option is disabled as it would result in a default allow
posture during agent downtime.
    
Also, only "match all" wildcard rules (i.e. `matchPattern = "*"`) are
unloaded for security reasons. Otherwise we could invalidate the L7
security rules.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -582,8 +582,76 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+				if len(portProtocols) != 1 || portProtocols[0].Port != "53" {
+					// not a L7 DNS rule
+					return nil
+				}
```
Arguably it's a bit more work (ie no early exit on the iteration), but I think we could iterate all of the rules and look for `len(portProtocols) > 0` + `len(pr.GetPortRule().Rules.DNS) > 0` and always apply the logic. IIRC we do support customising the ports used for DNS and some users remap the DNS ports. So while 'port 53' will catch most cases, we can probably make this a little bit more robust by not assuming this and only relying on the presence of DNS L7 rules in the policies.

Also, users can set up multiple port rules where the second or later is the DNS port rule. In that case, this logic wouldn't yet catch their case, even if it has L7 DNS rules. So I think it's a bit more robust to just iterate through all of the portrules to make these checks.

tklauser replied:
Good point, I didn't consider that. Changed according to your suggestion.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -582,8 +582,76 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+
+			// Bump revision to trigger policy recalculation
+			log.Info("Triggering policy recalculation")
+			policyRepo.BumpRevision()
```
nit: I wonder if the revision should be bumped while holding the lock in the first place, not sure I fully understand the implications of making changes without bumping revision + later bumping.

tklauser replied:
It seems other call sites bumping the revision don't hold the lock either and looking at the [implementation of `policy.(*Repository).BumpRevision`](https://github.com/cilium/cilium/blob/4b49be463f1d9f8a524ce6507e18c25187786bcf/pkg/policy/repository.go#L624-L628) it looks like internally the revision is incremented atomically, so I think it is fine to bump without holding the lock.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -582,8 +582,76 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+			policyRepo.BumpRevision()
+			regenerationMetadata := &regeneration.ExternalRegenerationMetadata{
+				Reason:            "unloading DNS policies on graceful shutdown",
+				RegenerationLevel: regeneration.RegenerateWithDatapathRewrite,
```
I believe that `RegenerateWithoutDatapath` should be sufficient for this use case, as "reload" and "rebuild" of the datapath is not required. I think that the regeneration levels are not sufficiently documented to justify this, but my understanding is that `RegenerateWithoutDatapath` will do policy recalculation + bpf policymap population. Lowering this value will allow us to skip some of the more expensive steps like compiling BPF programs or loading BPF programs into the kernel (which should not be necessary to implement this desired change).

Worth double-checking; worst case we _can_ use a higher regeneration level like this, it's just more expensive (possibly adding seconds onto shutdown time). But these regeneration levels are intended to allow policy regeneration like this without some of those more expensive steps.

tklauser replied:
Changed it to `RegenerateWithoutDatapath` and the effect is still the same in my tests.

I think I must have blindly copied `regenerationMetadata` with `RegenerateWithDatapathReload` from somewhere else and forgot to review/change its field values :facepalm:

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -582,8 +582,76 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+			policyRepo.Mutex.Unlock()
+
+			// Bump revision to trigger policy recalculation
+			log.Info("Triggering policy recalculation")
```
nit: Logs like this should ideally be even more specific, not just "triggering policy recalculation" (which can be triggered from many places for many reasons), but `"Triggering policy recalculation to remove DNS rules due to --%s", option.DNSPolicyUnloadOnShutdown`.

joestringer replied:
(Probably same sort of feedback for the later logs)

tklauser replied:
Thanks, changed to more specific log messages.

aanm commented on daemon/cmd/daemon.go:
```diff
@@ -582,8 +582,76 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+				for _, er := range rule.Egress {
+					_ = er.ToPorts.Iterate(removeL7DNSRules)
+				}
+				// TODO: modify Ingress, IngressDeny and EgressDeny as well?
```
@joestringer do we need to modify ingress? I don't think so.

@tklauser regarding the [Ingress|Egress]Deny, they should be left untouched.

joestringer replied:
Pretty sure we have a docs page somewhere that says that DNS only supports egress. Agree we shouldn't need to worry there. +1 on not touching Denies.

tklauser replied:
Thanks for clarifying, I've removed the comment.

aanm reviewed (commented):
I hadn't thought on stopping all watchers which is actually better than only stop the policy-related ones.

tklauser committed:
k8s: allow to stop k8s watchers

Introduce (*K8sWatcher).stop channel which is closed when
(*K8sWatcher).StopK8sServiceHandler is called, causing all k8s object
watchers to stop.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
policy: add a method to iterate all rules in the repository

This will be used in successive commits to modify rules on graceful
shutdown.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
daemon: allow running pre-cleanup functions

This allows to add functions to be run before the daemon context is
cancelled and the actual cleanup functions are processed in the daemon
signal handler.

This will be used in a successive commit to unload DNS policy rules and
regenerate the endpoints while all related subsystems are still
functional.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

sayboras commented on daemon/cmd/daemon.go:
```diff
@@ -582,8 +582,73 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+			policyRepo.Mutex.Unlock()
+
+			// Bump revision to trigger policy recalculation
+			log.Infof("Triggering policy recalculation to remove DNS rules due to --%s", option.DNSPolicyUnloadOnShutdown)
```
i am not sure how costly this regeneration is, what do think about something like below ?

```
		regenRequired := false
		cleaner.preCleanupFuncs.Add(func() {
		...
				if hasWildcard {
					portRule.Rules = nil
					regenRequired = true
				}
				return nil
                ...
                }
                ...
			if !regenRequired {
				log.Infof("Happy day, no ops due to due to --%s", option.DNSPolicyUnloadOnShutdown)
				return
			}
```

tklauser replied:
I'm not sure I completely understand the suggestion. The regeneration needs to happen inside the pre-cleanup func, so it is performed before shutting down the agent. Also we want to make sure to iterate all possible rules. The way I read your suggestion this wouldn't be the case.

sayboras replied:
oh sorry, probably the below diff will make it clearer

<details>

```diff
diff --git a/daemon/cmd/daemon.go b/daemon/cmd/daemon.go
index 1490322014..6d49d7809a 100644
--- a/daemon/cmd/daemon.go
+++ b/daemon/cmd/daemon.go
@@ -593,6 +593,8 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 			log.Info("Stopping k8s service handler")
 			d.k8sWatcher.StopK8sServiceHandler()
 
+			regenRequired := false
+
 			// Iterate over the policy repository and remove L7 DNS part
 			removeL7DNSRules := func(pr policyAPI.Ports) error {
 				portProtocols := pr.GetPortProtocols()
@@ -621,6 +623,7 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 				}
 				if hasWildcard {
 					portRule.Rules = nil
+					regenRequired = true
 				}
 				return nil
 			}
@@ -634,6 +637,11 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 			})
 			policyRepo.Mutex.Unlock()
 
+			if !regenRequired {
+				log.Infof("Happy day, no ops due to due to --%s", option.DNSPolicyUnloadOnShutdown)
+				return
+			}
+
 			// Bump revision to trigger policy recalculation
 			log.Infof("Triggering policy recalculation to remove DNS rules due to --%s", option.DNSPolicyUnloadOnShutdown)
 			policyRepo.BumpRevision()

```

</details>

tklauser replied:
Thanks for the diff @sayboras! Agree, there is no need to trigger a policy regeneration in case of no DNS rules. Will change accordingly.

sayboras reviewed (approved):
Small comment as per below, LGTM :100:

tklauser committed:
daemon: allow unloading DNS policy rules on graceful shutdown

Introduce the --dns-policy-unload-on-shutdown option which will cause
DNS policy rules to be unloaded on graceful agent shutdown.

This leads to DNS queries no longer being redirected to the L7 DNS proxy
by the BPF datapath and allows endpoints on a node where the agent is
not running to continue being able to resolve DNS queries.

By default the option is disabled as it would result in a default allow
posture during agent downtime.

Also, only "match all" wildcard rules (i.e. matchPattern = "*") are
unloaded for security reasons. Otherwise we could invalidate the L7
security rules.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
install/kubernetes: add values for setting termination grace period

This allows to set the termination grace period for agent, preflight and
hubble-relay DaemonSets/Deployments. Existing defaults are kept, but in
some cases (e.g. when unloading DNS L7 policy rules on graceful
shutdown), users might want to increase the grace period.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

joestringer reviewed (approved):
LGTM.

sayboras commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -70,6 +70,15 @@ serviceAccounts:
     annotations: {}
 
+# -- Configure termination grace period for components.
+terminationGracePeriods:
```
I feel like this attribute should go to respective components to be closed to all Deployment attributes. Understand that we have ServiceAccount section for all components in top level, but SA seems to be special case.

.Values.terminationGracePeriodSeconds --> agent
.Values.preflight.terminationGracePeriodSeconds -> preflight
.Values.relay.terminationGracePeriodSeconds -> relay

ldelossa commented on pkg/policy/repository.go:
```diff
@@ -420,6 +420,13 @@ func (p *Repository) AddList(rules api.Rules) (ruleSlice, uint64) {
 }
 
+// Iterate iterates the policy repository, calling f for each rule.
+func (p *Repository) IterateLocked(f func(rule *api.Rule)) {
```
Comment seems to be directed at the wrong function

aanm replied:
Fixed

ldelossa commented on pkg/policy/repository_test.go:
```diff
@@ -2118,3 +2118,88 @@ func (ds *PolicyTestSuite) TestremoveIdentityFromRuleCaches(c *C) {
+		}
+	}
+	repo.Mutex.Lock()
+	repo.IterateLocked(countEgressRules)
```
who's accessing this concurrently? Looks like you create it above. Is `NewPolicyRepository` spawning some go-routines?

aanm replied:
Right now, nothing, but in the future we might have something accessing concurrently.

ldelossa commented on daemon/cmd/daemon.go:
```diff
@@ -582,8 +582,80 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+
+			policyRepo := d.GetPolicyRepository()
+			policyRepo.Mutex.Lock()
+			policyRepo.IterateLocked(func(rule *policyAPI.Rule) {
```
One would expect `IterateLocked` would lock and unlock the `policyRepo` for the caller. Why make the caller do it?

joestringer replied:
In Cilium we read this roughly as `IterateWhileLocked()`: The idea is that everything should be more-or-less 'safe' when using the regular version (eg `Iterate()`). But if you need to hold the lock across multiple operations for some reason, then expose a special locked version that informs the caller that they need to be considering the correct locking to ensure the safety.

ldelossa replied:
Is it not more effective to have `IterateLocked` do the locking? or is the idea that the caller may want to perform some other actions once the tight loop returns?

joestringer replied:
Usually when we expose functions like `IterateLocked()`, the caller wants to do other operations or do multiple operations while holding the lock. Looking at this case though, I think it would be OK to move the locking inside and rename `IterateLocked()` -> `Iterate()`.

ldelossa replied:
Cool, I agree on both points. The semantics of `IterateLocked` is nice. And the usage here does appear locking can be internal and a bit safer for the caller.

aanm replied:
Agreed, fixed. IterateLocked is now Iterate and the mutex is moved inside.

ldelossa reviewed (changes requested):
Some questions.

sayboras commented on pkg/policy/repository_test.go:
```diff
@@ -2118,3 +2118,80 @@ func (ds *PolicyTestSuite) TestremoveIdentityFromRuleCaches(c *C) {
+	}
+	repo.Mutex.Lock()
+	repo.Iterate(countEgressRules)
+	repo.Mutex.Unlock()
```
```suggestion
 	repo.Iterate(countEgressRules)
```
